tests/functional/deps/test_deps_with_vars.py

```diff
@@ -0,0 +1,343 @@
+"""Test that dbt deps works when vars are used in dbt_project.yml without defaults.
+
+The key behavior being tested:
+- dbt deps uses lenient mode (require_vars=False) and succeeds even with missing vars
+- dbt run/compile/build/debug use strict mode (require_vars=True) and show the right error messages
+
+Expected behavior from reviewer's scenario:
+1. dbt deps succeeds (doesn't need vars)
+2. dbt run fails with error "Required var 'X' not found"
+3. dbt run --vars succeeds when vars provided
+"""
+
+import os
+
+import pytest
+import yaml
+
+from dbt.tests.util import run_dbt
+from dbt_common.exceptions import CompilationError
+
+# Simple model for testing
+model_sql = """
+select 1 as id
+"""
+
+
```

Comment:
A majority of these tests rely on essentially the same fixtures. As such it'd make sense to create a base test class that the actual test class inherit. Something like
```
class VarTestingBase:
    @pytest.fixture(scope="class")
    def models(self):
        return {"test_model.sql": model_sql}

    @pytest.fixture(scope="class")
    def project_config_update(self):
        return {
            "models": {
                "test_project": {"+materialized": "{{ var('materialized_var', 'view') }}"}
            }
        }
```

Two test classes would need to override the `project_config_update` fixture such that it doesn't contain the default value for the var. Those test classes are
* `TestCompileFailsWithMissingVar`
* `TestBuildFailsWithMissingVar`